File: server/sensor_fusion/diagnostics.py
from __future__ import annotations

import secrets

from .stream_schema import EvidenceEntity, SensorEvent, utc_now


def make_id(prefix: str) -> str:
    return f"{prefix}:{secrets.token_hex(6)}"


def location_label(entity: EvidenceEntity) -> str | None:
    if entity.lat is not None and entity.lon is not None:
        return f"{entity.lat:.3f}, {entity.lon:.3f}"
    return None


def event_from_entity(event_type: str, entity: EvidenceEntity, summary: str) -> SensorEvent:
    return SensorEvent(
        id=make_id(event_type),
        type=event_type,
        entity_id=entity.id,
        provider=entity.provider,
        title=entity.title,
        summary=summary,
        severity=entity.severity or "info",
        confidence=entity.confidence,
        created_at=utc_now(),
        location=location_label(entity),
        payload={"entity": entity.__dict__},
    )

File: src/server/sensorFusion/diagnostics.ts
import type { EvidenceEntity, EvidenceSeverity, SensorEvent } from "./streamSchema";

export function nowIso(): string {
    return new Date().toISOString();
}

export function makeId(prefix: string): string {
    return `${prefix}:${Date.now().toString(36)}:${Math.random().toString(36).slice(2, 8)}`;
}

export function locationLabel(entity: Pick<EvidenceEntity, "lat" | "lon" | "raw">): string | null {
    const city = typeof entity.raw.city === "string" ? entity.raw.city : null;
    const state = typeof entity.raw.state === "string" ? entity.raw.state : null;
    const country = typeof entity.raw.country === "string" ? entity.raw.country : null;
    const named = [city, state, country].filter(Boolean).join(", ");
    if (named) return named;
    if (entity.lat !== null && entity.lon !== null) return `${entity.lat.toFixed(3)}, ${entity.lon.toFixed(3)}`;
    return null;
}

export function eventFromEntity(type: SensorEvent["type"], entity: EvidenceEntity, summary: string, severity: EvidenceSeverity = entity.severity ?? "info"): SensorEvent {
    return {
        id: makeId(type),
        type,
        entity_id: entity.id,
        provider: entity.provider,
        severity,
        confidence: entity.confidence,
        title: entity.title,
        summary,
        location: locationLabel(entity),
        created_at: nowIso(),
        payload: { entity },
    };
}

export function staleWarning(staleCount: number): string | null {
    if (staleCount <= 0) return null;
    return `${staleCount.toLocaleString()} sensor entities are stale or past expected refresh. Treat conclusions as degraded until sources recover.`;
}
